File: riveredge-backend/src/apps/haoligo/constants/route_patrol_line_status.py
"""好力 GO — 路线巡检行状态。"""

from __future__ import annotations

from typing import Any

ROUTE_PATROL_LINE_STATUS_NORMAL = "normal"
ROUTE_PATROL_LINE_STATUS_ABNORMAL = "abnormal"
ROUTE_PATROL_LINE_STATUS_NOT_PRODUCING = "not_producing"

ROUTE_PATROL_LINE_STATUSES = frozenset(
    {
        ROUTE_PATROL_LINE_STATUS_NORMAL,
        ROUTE_PATROL_LINE_STATUS_ABNORMAL,
        ROUTE_PATROL_LINE_STATUS_NOT_PRODUCING,
    }
)


def normalize_route_patrol_line_status(value: Any, *, default: str = ROUTE_PATROL_LINE_STATUS_NORMAL) -> str:
    raw = str(value or "").strip().lower()
    if raw in ROUTE_PATROL_LINE_STATUSES:
        return raw
    if raw in {"true", "1", "yes"}:
        return ROUTE_PATROL_LINE_STATUS_NORMAL
    if raw in {"false", "0", "no"}:
        return ROUTE_PATROL_LINE_STATUS_ABNORMAL
    return default


def is_route_patrol_line_abnormal(status: str | None) -> bool:
    return normalize_route_patrol_line_status(status) == ROUTE_PATROL_LINE_STATUS_ABNORMAL

File: riveredge-frontend/src/apps/haoligo/constants/routePatrolLineStatus.ts
export const ROUTE_PATROL_LINE_STATUS_NORMAL = 'normal' as const;
export const ROUTE_PATROL_LINE_STATUS_ABNORMAL = 'abnormal' as const;
export const ROUTE_PATROL_LINE_STATUS_NOT_PRODUCING = 'not_producing' as const;

export type RoutePatrolLineStatus =
  | typeof ROUTE_PATROL_LINE_STATUS_NORMAL
  | typeof ROUTE_PATROL_LINE_STATUS_ABNORMAL
  | typeof ROUTE_PATROL_LINE_STATUS_NOT_PRODUCING;

export const ROUTE_PATROL_LINE_STATUS_OPTIONS: Array<{
  value: RoutePatrolLineStatus;
  labelKey: string;
  fallback: string;
}> = [
  {
    value: ROUTE_PATROL_LINE_STATUS_NORMAL,
    labelKey: 'app.haoligo.equipment.documents.resultNormal',
    fallback: '正常',
  },
  {
    value: ROUTE_PATROL_LINE_STATUS_ABNORMAL,
    labelKey: 'app.haoligo.equipment.documents.resultAbnormal',
    fallback: '异常',
  },
  {
    value: ROUTE_PATROL_LINE_STATUS_NOT_PRODUCING,
    labelKey: 'app.haoligo.equipment.documents.resultNotProducing',
    fallback: '未生产',
  },
];

export function normalizeRoutePatrolLineStatus(
  value: unknown,
  fallback: RoutePatrolLineStatus = ROUTE_PATROL_LINE_STATUS_NORMAL,
): RoutePatrolLineStatus {
  const raw = String(value ?? '').trim().toLowerCase();
  if (raw === ROUTE_PATROL_LINE_STATUS_ABNORMAL) return ROUTE_PATROL_LINE_STATUS_ABNORMAL;
  if (raw === ROUTE_PATROL_LINE_STATUS_NOT_PRODUCING) return ROUTE_PATROL_LINE_STATUS_NOT_PRODUCING;
  if (raw === ROUTE_PATROL_LINE_STATUS_NORMAL) return ROUTE_PATROL_LINE_STATUS_NORMAL;
  if (raw === 'false') return ROUTE_PATROL_LINE_STATUS_ABNORMAL;
  if (raw === 'true') return ROUTE_PATROL_LINE_STATUS_NORMAL;
  return fallback;
}

export function isRoutePatrolLineAbnormal(status: unknown): boolean {
  return normalizeRoutePatrolLineStatus(status) === ROUTE_PATROL_LINE_STATUS_ABNORMAL;
}

export function patchRoutePatrolLineStatus<T extends { line_status: RoutePatrolLineStatus; abnormal_description?: string | null; attachment_file_ids?: string[] | null }>(
  line: T,
  nextStatus: RoutePatrolLineStatus,
): T {
  if (nextStatus === ROUTE_PATROL_LINE_STATUS_ABNORMAL) {
    return { ...line, line_status: nextStatus };
  }
  return {
    ...line,
    line_status: nextStatus,
    abnormal_description: null,
    attachment_file_ids: null,
  };
}
